sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/write/DocumentWriteService.java

```diff
@@ -0,0 +1,115 @@
+import io.stargate.bridge.proto.QueryOuterClass.ResultSet;
+import io.stargate.bridge.proto.StargateBridge;
+import io.stargate.sgv2.docsapi.api.common.StargateRequestInfo;
+import io.stargate.sgv2.docsapi.api.common.properties.datastore.DataStoreProperties;
+import io.stargate.sgv2.docsapi.api.common.properties.document.DocumentProperties;
+import io.stargate.sgv2.docsapi.service.ExecutionContext;
+import io.stargate.sgv2.docsapi.service.JsonShreddedRow;
+import io.stargate.sgv2.docsapi.service.util.TimeSource;
+import io.stargate.sgv2.docsapi.service.write.db.InsertQueryBuilder;
+import java.util.List;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+@ApplicationScoped
+public class DocumentWriteService {
+
+  private final StargateRequestInfo requestInfo;
+  private final TimeSource timeSource;
+  private final InsertQueryBuilder insertQueryBuilder;
+  private final boolean useLoggedBatches;
+
+  @Inject
+  public DocumentWriteService(
+      StargateRequestInfo requestInfo,
+      TimeSource timeSource,
+      DataStoreProperties dataStoreProperties,
+      DocumentProperties documentProperties) {
+    this.requestInfo = requestInfo;
+    this.insertQueryBuilder = new InsertQueryBuilder(documentProperties);
+    this.timeSource = timeSource;
+    this.useLoggedBatches = dataStoreProperties.loggedBatchesEnabled();
+  }
+
+  /**
+   * Writes a single document, without deleting the existing document with the same key. Please call
+   * this method only if you are sure that the document with given ID does not exist.
+   *
+   * @param keyspace Keyspace to store document in.
+   * @param collection Collection the document belongs to.
+   * @param documentId Document ID.
+   * @param rows Rows of this document.
+   * @param ttl the time-to-live of the rows (seconds)
+   * @param numericBooleans If numeric boolean should be stored.
+   * @param context Execution content for profiling.
+   * @return Single containing the {@link ResultSet} of the batch execution.
+   */
+  public Uni<ResultSet> writeDocument(
+      String keyspace,
+      String collection,
+      String documentId,
+      List<JsonShreddedRow> rows,
+      Integer ttl,
+      boolean numericBooleans,
+      ExecutionContext context) {
+
+    StargateBridge bridge = requestInfo.getStargateBridge();
+
+    return Uni.createFrom()
+        .item(() -> insertQueryBuilder.buildQuery(keyspace, collection, ttl))
+        .map(
+            query -> {
+              long timestamp = timeSource.currentTimeMicros();
+              return rows.stream()
+                  .map(
+                      row ->
+                          insertQueryBuilder.bind(
+                              query, documentId, row, timestamp, numericBooleans))
+                  .toList();
+            })
+        .flatMap(boundQueries -> executeBatch(bridge, boundQueries, context.nested("ASYNC INSERT")))
+    // TODO reschedule outside of gRPC executor?
```

Comment:
What is the threading model in v2? Do we want to reschedule this on a different executor?

I'm not sure if that's really necessary, unless the caller chains really compute-intensive work on it. Context switching has a cost too.